Flaky unit test: vreplication

```
ok  	vitess.io/vitess/go/vt/vttablet/tabletmanager	1.164s
E1206 23:43:25.105640   16539 binlog_player.go:202] error (expected error) in selecting vreplication settings select pos, stop_pos, max_tps, max_replication_lag, state from _vt.vreplication where id=1
E1206 23:43:25.107048   16539 controller.go:149] stream 1: error (expected error) in selecting vreplication settings select pos, stop_pos, max_tps, max_replication_lag, state from _vt.vreplication where id=1, retrying after 10ms
E1206 23:43:25.446839   16539 engine.go:596] Error waiting for pos: MariaDB/0-1-1084, last pos: MariaDB/0-1-1083: context canceled, wait time: 45.9µs
--- FAIL: TestPlayerCopyWildcardTableContinuation (2.99s)
##[error]    vcopier_test.go:613: query:
        "update _vt.vreplication set pos='MySQL56/2f7cf4b4-1882-11ea-a57c-000d3a0454a7:1-179', time_updated=1575675813, transaction_timestamp=1575675811, message='' where id=14", does not match query 0:
        "/update _vt.vreplication set state='Running'"
E1206 23:43:36.056591   16539 controller.go:149] stream 28: Duplicate column name 'val2' (errno 1060) (sqlstate 42S21) during query: alter table t1 add column val2 varchar(128), retrying after 5s
E1206 23:43:43.314132   16539 controller.go:149] stream 38: vstream ended, retrying after 1ms
E1206 23:43:43.658767   16539 vstreamer.go:648] stream error @ 2f7cf4b4-1882-11ea-a57c-000d3a0454a7:1-593: unknown table t1 in schema
FAIL
FAIL	vitess.io/vitess/go/vt/vttablet/tabletmanager/vreplication	35.409s
```

--- a/.github/workflows/unit.yml
+++ b/.github/workflows/unit.yml
@@ -6,7 +6,7 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        name: [mysql57, mysql80]
+        name: [percona56, mysql57, mysql80, mariadb101, mariadb102, mariadb103]
 
     steps:
     - name: Set up Go

--- a/go.mod
+++ b/go.mod
@@ -25,7 +25,6 @@ require (
 	github.com/golang/mock v1.3.1
 	github.com/golang/protobuf v1.3.2
 	github.com/golang/snappy v0.0.0-20170215233205-553a64147049
-	github.com/google/btree v1.0.0 // indirect
 	github.com/google/shlex v0.0.0-20181106134648-c34317bd91bf // indirect
 	github.com/gorilla/websocket v0.0.0-20160912153041-2d1e4548da23
 	github.com/grpc-ecosystem/go-grpc-middleware v1.1.0
@@ -50,8 +49,6 @@ require (
 	github.com/minio/minio-go v0.0.0-20190131015406-c8a261de75c1
 	github.com/mitchellh/go-testing-interface v1.0.0 // indirect
 	github.com/mitchellh/mapstructure v1.1.2 // indirect
-	github.com/modern-go/concurrent v0.0.0-20180306012644-bacd9c7ef1dd // indirect
-	github.com/modern-go/reflect2 v1.0.1 // indirect
 	github.com/olekukonko/tablewriter v0.0.0-20160115111002-cca8bbc07984
 	github.com/opentracing-contrib/go-grpc v0.0.0-20180928155321-4b5a12d3ff02
 	github.com/opentracing/opentracing-go v1.1.0
@@ -60,7 +57,7 @@ require (
 	github.com/pkg/errors v0.8.1
 	github.com/prometheus/client_golang v1.1.0
 	github.com/prometheus/client_model v0.0.0-20190812154241-14fe0d1b01d4 // indirect
-	github.com/prometheus/common v0.7.0 // indirect
+	github.com/prometheus/common v0.7.0
 	github.com/prometheus/procfs v0.0.5 // indirect
 	github.com/satori/go.uuid v0.0.0-20160713180306-0aa62d5ddceb // indirect
 	github.com/stretchr/testify v1.4.0

--- a/go/mysql/client.go
+++ b/go/mysql/client.go
@@ -380,7 +380,7 @@ func (c *Conn) parseInitialHandshakePacket(data []byte) (uint32, []byte, error) 
 		errorCode, pos, _ := readUint16(data, pos)
 		// Normally there would be a 1-byte sql_state_marker field and a 5-byte
 		// sql_state field here, but docs say these will not be present in this case.
-		errorMsg, pos, _ := readEOFString(data, pos)
+		errorMsg, _, _ := readEOFString(data, pos)
 		return 0, nil, NewSQLError(CRServerHandshakeErr, SSUnknownSQLState, "immediate error from server errorCode=%v errorMsg=%v", errorCode, errorMsg)
 	}
 

--- a/go/mysql/flavor_filepos.go
+++ b/go/mysql/flavor_filepos.go
@@ -21,6 +21,7 @@ import (
 	"fmt"
 	"io"
 	"strconv"
+	"strings"
 	"time"
 
 	"golang.org/x/net/context"
@@ -143,10 +144,17 @@ func (flv *filePosFlavor) readBinlogEvent(c *Conn) (BinlogEvent, error) {
 				// No need to transmit. Just update the internal position for the next event.
 				continue
 			}
-		case eXIDEvent, eQueryEvent, eTableMapEvent,
+		case eXIDEvent, eTableMapEvent,
 			eWriteRowsEventV0, eWriteRowsEventV1, eWriteRowsEventV2,
 			eDeleteRowsEventV0, eDeleteRowsEventV1, eDeleteRowsEventV2,
 			eUpdateRowsEventV0, eUpdateRowsEventV1, eUpdateRowsEventV2:
+			flv.savedEvent = event
+			return newFilePosGTIDEvent(flv.file, event.nextPosition(flv.format), event.Timestamp()), nil
+		case eQueryEvent:
+			q, err := event.Query(flv.format)
+			if err == nil && strings.HasPrefix(q.SQL, "#") {
+				continue
+			}
 			flv.savedEvent = event
 			return newFilePosGTIDEvent(flv.file, event.nextPosition(flv.format), event.Timestamp()), nil
 		default:

--- a/go/mysql/server_test.go
+++ b/go/mysql/server_test.go
@@ -719,11 +719,6 @@ func TestServer(t *testing.T) {
 // TestClearTextServer creates a Server that needs clear text
 // passwords from the client.
 func TestClearTextServer(t *testing.T) {
-	// If the database we're using is MariaDB, the client
-	// is also the MariaDB client, that does support
-	// clear text by default.
-	isMariaDB := os.Getenv("MYSQL_FLAVOR") == "MariaDB"
-
 	th := &testHandler{}
 
 	authServer := NewAuthServerStatic()
@@ -740,6 +735,9 @@ func TestClearTextServer(t *testing.T) {
 	go l.Accept()
 
 	host, port := getHostPort(t, l.Addr())
+
+	version, _ := runMysql(t, nil, "--version")
+	isMariaDB := strings.Contains(version, "MariaDB")
 
 	// Setup the right parameters.
 	params := &ConnParams{
@@ -1170,30 +1168,34 @@ func runMysql(t *testing.T, params *ConnParams, command string) (string, bool) {
 		command = command[len(enableCleartextPluginPrefix):]
 		args = append(args, "--enable-cleartext-plugin")
 	}
-	args = append(args, "-e", command)
-	if params.UnixSocket != "" {
-		args = append(args, "-S", params.UnixSocket)
+	if command == "--version" {
+		args = append(args, command)
 	} else {
-		args = append(args,
-			"-h", params.Host,
-			"-P", fmt.Sprintf("%v", params.Port))
-	}
-	if params.Uname != "" {
-		args = append(args, "-u", params.Uname)
-	}
-	if params.Pass != "" {
-		args = append(args, "-p"+params.Pass)
-	}
-	if params.DbName != "" {
-		args = append(args, "-D", params.DbName)
-	}
-	if params.Flags&CapabilityClientSSL > 0 {
-		args = append(args,
-			"--ssl",
-			"--ssl-ca", params.SslCa,
-			"--ssl-cert", params.SslCert,
-			"--ssl-key", params.SslKey,
-			"--ssl-verify-server-cert")
+		args = append(args, "-e", command)
+		if params.UnixSocket != "" {
+			args = append(args, "-S", params.UnixSocket)
+		} else {
+			args = append(args,
+				"-h", params.Host,
+				"-P", fmt.Sprintf("%v", params.Port))
+		}
+		if params.Uname != "" {
+			args = append(args, "-u", params.Uname)
+		}
+		if params.Pass != "" {
+			args = append(args, "-p"+params.Pass)
+		}
+		if params.DbName != "" {
+			args = append(args, "-D", params.DbName)
+		}
+		if params.Flags&CapabilityClientSSL > 0 {
+			args = append(args,
+				"--ssl",
+				"--ssl-ca", params.SslCa,
+				"--ssl-cert", params.SslCert,
+				"--ssl-key", params.SslKey,
+				"--ssl-verify-server-cert")
+		}
 	}
 	env := []string{
 		"LD_LIBRARY_PATH=" + path.Join(dir, "lib/mysql"),

--- a/go/vt/vttablet/filelogger/filelogger_test.go
+++ b/go/vt/vttablet/filelogger/filelogger_test.go
@@ -60,13 +60,19 @@ func TestFileLog(t *testing.T) {
 	tabletenv.StatsLogger.Send(log2)
 
 	// Allow time for propagation
-	time.Sleep(10 * time.Millisecond)
+	for i := 0; i < 10; i++ {
+		time.Sleep(10 * time.Millisecond)
 
-	want := "\t\t\t''\t''\t0001-01-01 00:00:00.000000\t0001-01-01 00:00:00.000000\t0.000000\t\t\"test 1\"\tmap[]\t1\t\"test 1 PII\"\tmysql\t0.000000\t0.000000\t0\t0\t\"\"\t\n\t\t\t''\t''\t0001-01-01 00:00:00.000000\t0001-01-01 00:00:00.000000\t0.000000\t\t\"test 2\"\tmap[]\t1\t\"test 2 PII\"\tmysql\t0.000000\t0.000000\t0\t0\t\"\"\t\n"
-	contents, _ := ioutil.ReadFile(logPath)
-	got := string(contents)
-	if want != string(got) {
-		t.Errorf("streamlog file: want %q got %q", want, got)
+		want := "\t\t\t''\t''\t0001-01-01 00:00:00.000000\t0001-01-01 00:00:00.000000\t0.000000\t\t\"test 1\"\tmap[]\t1\t\"test 1 PII\"\tmysql\t0.000000\t0.000000\t0\t0\t\"\"\t\n\t\t\t''\t''\t0001-01-01 00:00:00.000000\t0001-01-01 00:00:00.000000\t0.000000\t\t\"test 2\"\tmap[]\t1\t\"test 2 PII\"\tmysql\t0.000000\t0.000000\t0\t0\t\"\"\t\n"
+		contents, _ := ioutil.ReadFile(logPath)
+		got := string(contents)
+		if want == got {
+			return
+		}
+		// Last iteration.
+		if i == 9 {
+			t.Errorf("streamlog file: want %q got %q", want, got)
+		}
 	}
 }
 

--- a/go/vt/vttablet/tabletmanager/vreplication/vstreamer_client_test.go
+++ b/go/vt/vttablet/tabletmanager/vreplication/vstreamer_client_test.go
@@ -184,6 +184,7 @@ func TestTabletVStreamerClientVStream(t *testing.T) {
 		"drop table t1",
 		fmt.Sprintf("drop table %s.t1", vrepldb),
 	})
+	env.SchemaEngine.Reload(context.Background())
 
 	ctx := context.Background()
 	err := vsClient.Open(ctx)
@@ -241,6 +242,7 @@ func TestTabletVStreamerClientVStreamRows(t *testing.T) {
 		"drop table t1",
 		fmt.Sprintf("drop table %s.t1", vrepldb),
 	})
+	env.SchemaEngine.Reload(context.Background())
 
 	qr, err := env.Mysqld.FetchSuperQuery(context.Background(), "select now()")
 	if err != nil {
@@ -443,6 +445,7 @@ func TestMySQLVStreamerClientVStream(t *testing.T) {
 		"drop table t1",
 		fmt.Sprintf("drop table %s.t1", vrepldb),
 	})
+	env.SchemaEngine.Reload(context.Background())
 
 	ctx := context.Background()
 	err := vsClient.Open(ctx)
@@ -497,6 +500,7 @@ func TestMySQLVStreamerClientVStreamRows(t *testing.T) {
 		"drop table t1",
 		fmt.Sprintf("drop table %s.t1", vrepldb),
 	})
+	env.SchemaEngine.Reload(context.Background())
 
 	qr, err := env.Mysqld.FetchSuperQuery(context.Background(), "select now()")
 	if err != nil {

--- a/go/vt/vttablet/tabletserver/vstreamer/main_test.go
+++ b/go/vt/vttablet/tabletserver/vstreamer/main_test.go
@@ -21,6 +21,7 @@ import (
 	"fmt"
 	"os"
 	"testing"
+	"time"
 
 	"vitess.io/vitess/go/vt/vttablet/tabletserver/vstreamer/testenv"
 )
@@ -52,6 +53,12 @@ func TestMain(m *testing.M) {
 		engine.InitDBConfig(env.Dbcfgs.DbaWithDB())
 		engine.Open(env.KeyspaceName, env.Cells[0])
 		defer engine.Close()
+
+		// GH actions machines sometimes exhibit multi-second delays on replication.
+		// So, give a generously high value for heartbeat time for all tests.
+		saveHeartbeat := HeartbeatTime
+		defer func() { HeartbeatTime = saveHeartbeat }()
+		HeartbeatTime = 10 * time.Second
 
 		return m.Run()
 	}()

--- a/go/vt/vttablet/tabletserver/vstreamer/vstreamer_test.go
+++ b/go/vt/vttablet/tabletserver/vstreamer/vstreamer_test.go
@@ -55,8 +55,6 @@ func TestStatements(t *testing.T) {
 			"update stream1 set val='bbb' where id = 1",
 			"commit",
 		},
-		// MySQL issues GTID->BEGIN.
-		// MariaDB issues BEGIN->GTID.
 		output: [][]string{{
 			`begin`,
 			`type:FIELD field_event:<table_name:"stream1" fields:<name:"id" type:INT32 > fields:<name:"val" type:VARBINARY > > `,
@@ -111,34 +109,6 @@ func TestStatements(t *testing.T) {
 			`gtid`,
 			`type:DDL ddl:"truncate table stream2" `,
 		}},
-	}, {
-		// repair, optimize and analyze show up in binlog stream, but ignored by vitess.
-		input: "repair table stream2",
-		output: [][]string{{
-			`gtid`,
-			`type:OTHER `,
-		}},
-	}, {
-		input: "optimize table stream2",
-		output: [][]string{{
-			`gtid`,
-			`type:OTHER `,
-		}},
-	}, {
-		input: "analyze table stream2",
-		output: [][]string{{
-			`gtid`,
-			`type:OTHER `,
-		}},
-	}, {
-		// select, set, show and describe don't get logged.
-		input: "select * from stream1",
-	}, {
-		input: "set @val=1",
-	}, {
-		input: "show tables",
-	}, {
-		input: "describe stream1",
 	}}
 	runCases(t, nil, testcases, "")
 
@@ -146,6 +116,70 @@ func TestStatements(t *testing.T) {
 	engine.cp.Flavor = "FilePos"
 	defer func() { engine.cp.Flavor = "" }()
 	runCases(t, nil, testcases, "")
+}
+
+// TestOther tests "other" statements. These statements produce
+// very different events depending on the version of mysql or mariadb
+// So, we just show that vreplication transmits "OTHER" events
+// if the binlog is affected by the statement.
+func TestOther(t *testing.T) {
+	if testing.Short() {
+		t.Skip()
+	}
+
+	execStatements(t, []string{
+		"create table stream1(id int, val varbinary(128), primary key(id))",
+		"create table stream2(id int, val varbinary(128), primary key(id))",
+	})
+	defer execStatements(t, []string{
+		"drop table stream1",
+		"drop table stream2",
+	})
+	engine.se.Reload(context.Background())
+
+	testcases := []string{
+		"repair table stream2",
+		"optimize table stream2",
+		"analyze table stream2",
+		"select * from stream1",
+		"set @val=1",
+		"show tables",
+		"describe stream1",
+	}
+
+	// customRun is a modified version of runCases.
+	customRun := func(mode string) {
+		t.Logf("Run mode: %v", mode)
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+		ch := startStream(ctx, t, nil, "")
+
+		want := [][]string{{
+			`gtid`,
+			`type:OTHER `,
+		}}
+
+		for _, stmt := range testcases {
+			startPosition := masterPosition(t)
+			execStatement(t, stmt)
+			endPosition := masterPosition(t)
+			if startPosition == endPosition {
+				t.Logf("statement %s did not affect binlog", stmt)
+				continue
+			}
+			expectLog(ctx, t, stmt, ch, want)
+		}
+		cancel()
+		if evs, ok := <-ch; ok {
+			t.Fatalf("unexpected evs: %v", evs)
+		}
+	}
+	customRun("gtid")
+
+	// Test FilePos flavor
+	engine.cp.Flavor = "FilePos"
+	defer func() { engine.cp.Flavor = "" }()
+	customRun("filePos")
 }
 
 func TestRegexp(t *testing.T) {
@@ -424,8 +458,6 @@ func TestSelectFilter(t *testing.T) {
 			"insert into t1 values (2, 4, 'aaa')",
 			"commit",
 		},
-		// MySQL issues GTID->BEGIN.
-		// MariaDB issues BEGIN->GTID.
 		output: [][]string{{
 			`begin`,
 			`type:FIELD field_event:<table_name:"t1" fields:<name:"id2" type:INT32 > fields:<name:"val" type:VARBINARY > > `,

--- a/test.go
+++ b/test.go
@@ -106,7 +106,7 @@ const (
 	configFileName = "test/config.json"
 
 	// List of flavors for which a bootstrap Docker image is available.
-	flavors = "mariadb,mysql56,mysql57,percona,percona57"
+	flavors = "mysql56,mysql57,mysql80,mariadb,mariadb103,percona,percona57,percona80"
 )
 
 // Config is the overall object serialized in test/config.json.

